notification/slack/actions.go

```diff
@@ -0,0 +1,179 @@
+package slack
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/slack-go/slack"
+	"github.com/target/goalert/alert"
+	"github.com/target/goalert/config"
+	"github.com/target/goalert/permission"
+)
+
+// Handler responds to API requests from Slack
+type Handler struct {
+	c Config
+}
+
+// Payload represents the relevant payload information sent from Slack
+type Payload struct {
+	ResponseURL string `json:"response_url"`
+	Actions     []Action
+	Channel     slack.Channel
+}
+
+// Action represents the information given from an action event within Slack
+// e.g. clicking to acknowledge an alert from slack
+type Action struct {
+	ActionID string `json:"action_id"`
+	ActionTS string `json:"action_ts"`
+	BlockID  string `json:"block_id"`
+	Value    string
+}
+
+// NewHandler creates a new Handler, registering API routes using chi.
+func NewHandler(c Config) *Handler {
+	return &Handler{c: c}
+}
+
+func httpErr(w http.ResponseWriter, err error) error {
+	http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+	panic(err)
+}
+
+// validRequest is used to validate a request from Slack.
+// If the request is validated true is returned, false otherwise.
+// https://api.slack.com/authentication/verifying-requests-from-slack
+func validRequest(w http.ResponseWriter, req *http.Request) error {
+	if req.Method != "POST" {
+		return httpErr(w, errors.New("not a post"))
+	}
+
+	secret := config.FromContext(req.Context()).Slack.SigningSecret
+	sv, err := slack.NewSecretsVerifier(req.Header, secret)
+	if err != nil {
+		return err
+	}
+
+	defer req.Body.Close()
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return httpErr(w, err)
+	}
+	req.Body.Close()
+	req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+	_, err = sv.Write(body)
+	if err != nil {
+		return httpErr(w, err)
+	}
```

Comment:
Use `io.TeeReader` to update `sv` as we read the body. Also we can remove the `req.Body.Close` since it's redundant to the `defer` above.

```suggestion
	body, err := ioutil.ReadAll(io.TeeReader(req.Body, sv))
	if err != nil {
		return httpErr(w, err)
	}
	req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
```